### PR TITLE
Change exercise solutions and regenerate exercise instructions

### LINKED ISSUE #418: Modify exercise solutions and regenerate exercises
https://github.com/INRIA/scikit-learn-mooc/pull/408 has been merged and was tackling only the first module, 6 more modules to go.

To work on this:
- chose any `python_scripts/*sol*.py`
- modify it following https://github.com/INRIA/scikit-learn-mooc/pull/408#issuecomment-884790196
- regenerate the exercises (`python_scripts/*ex*.py) from the solutions
  ```
  make exercises
  ```
- fix the problems that are likely to arise
- iterate

Solution files to modify:
- [x] python_scripts/cross_validation_sol_01.py  
- [x] python_scripts/cross_validation_sol_02.py 
- [x] python_scripts/ensemble_sol_01.py         
- [x] python_scripts/ensemble_sol_02.py         
- [x] python_scripts/ensemble_sol_03.py         
- [x] python_scripts/ensemble_sol_04.py         
- [x] python_scripts/ensemble_sol_05.py         
- [x] python_scripts/feature_selection_sol_01.py
- [ ] python_scripts/linear_models_sol_01.py    
- [ ] python_scripts/linear_models_sol_02.py    
- [ ] python_scripts/linear_models_sol_03.py    
- [ ] python_scripts/linear_models_sol_04.py    
- [ ] python_scripts/linear_models_sol_05.py    
- [x] python_scripts/metrics_sol_01.py          
- [ ] python_scripts/metrics_sol_02.py          
- [ ] python_scripts/parameter_tuning_sol_02.py 
- [ ] python_scripts/parameter_tuning_sol_03.py 
- [ ] python_scripts/trees_sol_01.py            
- [ ] python_scripts/trees_sol_02.py

### PR DESCRIPTION
**Pull request by ArturoAmorQ**
Fixes #418

Solution files to modify:
- [x] python_scripts/cross_validation_sol_01.py  
- [x] python_scripts/cross_validation_sol_02.py 
- [x] python_scripts/ensemble_sol_01.py         
- [x] python_scripts/ensemble_sol_02.py         
- [x] python_scripts/ensemble_sol_03.py         
- [x] python_scripts/ensemble_sol_04.py         
- [x] python_scripts/ensemble_sol_05.py         
- [x] python_scripts/feature_selection_sol_01.py
- [x] python_scripts/linear_models_sol_01.py (Loïc)   
- [x] python_scripts/linear_models_sol_02.py (Loïc)   
- [x] python_scripts/linear_models_sol_03.py (Loïc)    
- [x] python_scripts/linear_models_sol_04.py (Loïc)   
- [x] python_scripts/linear_models_sol_05.py (Loïc)    
- [x] python_scripts/metrics_sol_01.py          
- [x] python_scripts/metrics_sol_02.py          
- [x] python_scripts/parameter_tuning_sol_02.py (Olivier)
- [x] python_scripts/parameter_tuning_sol_03.py (Olivier)
- [x] python_scripts/trees_sol_01.py (Olivier)            
- [x] python_scripts/trees_sol_02.py (Olivier)

### CODE REVIEW

**Comment by lesteve** [comment]
I'll do the linear_models ones.

**Comment by ogrisel** [comment]
@ArturoAmorQ I think it's better not to address all of #418 in a single PR, this is why I edited the description to state "Toward #418" instead of "Fix #418" so that merging this PR does not close #418 by mistake.

File: python_scripts/ensemble_sol_03.py [hunk 1]
```diff
@@ -24,13 +24,14 @@
 # ```
 
 # %% [markdown]
-# Then, create an `AbaBoostRegressor`. Use the function
+# Then, create an `AdaBoostRegressor`. Use the function
```
Comment (ogrisel) [inline python_scripts/ensemble_sol_03.py:27]:
```suggestion
# Then, create an `AdaBoostRegressor` instance. Use the function
```

**Comment by lesteve** [comment]
I was the one who did the editing sorry I thought I would push directly in this PR but maybe that was too optimistic ...

**Comment by ogrisel** [comment]
> I was the one who did the editing sorry I thought I would push directly in this PR but maybe that was too optimistic ...

That's also an option. As you wish. Shall we re-rerun the script and included the fixes for the re-generated exercises as part of the PR itself?

**Review by ogrisel — COMMENTED** [review]
while we are at it:

**Comment by ogrisel** [comment]
I will work on `python_scripts/trees_sol_01.py` then and also push in this PR.

**Comment by lesteve** [comment]
> Shall we re-rerun the script and included the fixes for the re-generated exercises

I think we should, it may introduce jupytext header metadata changes (e.g. jupytext version) but I think it's fine. Also it is a good way to check that the exercises and solutions are in sync.

**Comment by lesteve** [comment]
I have put a checkbox list on the top comment to try to track the progress.

File: python_scripts/cross_validation_sol_01.py [hunk 2]
```diff
@@ -123,29 +126,30 @@
...
-# %% [markdown]
+# %% [markdown] tags=["solution"]
 # Looking at the curve, we can clearly identify the over-fitting regime of
 # the SVC classifier when `gamma > 1`.
 # The best setting is around `gamma = 1` while for `gamma < 1`,
 # it is not very clear if the classifier is under-fitting but the
 # testing score is worse than for `gamma = 1`.
 
-# %% [markdown]
+# %% [markdown]
```
Comment (ogrisel) [inline python_scripts/cross_validation_sol_01.py:136]:
```suggestion
# %% [markdown]
```

**Comment by lesteve** [comment]
I pushed a modification of the script to make it easier to generate a single exercise. This should make it a bit easier to work on this PR:

```
python build_tools/generate-exercise-from-solution.py python_scripts/linear_models_ex_01.py
```

**Comment by lesteve** [comment]
I had a look at the diff, and quickly browse the resulting HTML for the few files that had significant changes. I think this is good enough to merge.

If you spot some problems, please open separate issues.

**Pull request merged**